File: test-suite/swift/path_traversal_clean/RequestPathsSafe.swift
import Foundation

struct UploadedFile {
    let filename: String
    let data: Data
}

struct FileIO {
    func streamFile(at path: String) -> String {
        return path
    }
}

struct Request {
    let query: [String: String]
    let headers: [String: String]
    let url: URL
    let file: UploadedFile
    let fileio: FileIO
}

enum PathError: Error {
    case escapedRoot
}

let documentRoot = URL(fileURLWithPath: "/srv/app/files")

func safeUnderRoot(_ base: URL, _ requested: String) throws -> URL {
    let root = base.standardizedFileURL
    let candidate = root.appendingPathComponent(requested).standardizedFileURL
    guard candidate.path == root.path || candidate.path.hasPrefix(root.path + "/") else {
        throw PathError.escapedRoot
    }
    return candidate
}

func readDownload(req: Request) throws -> String {
    let requestedName = req.query["file"] ?? "index.html"
    let target = try safeUnderRoot(documentRoot, requestedName)
    return try String(contentsOf: target)
}

func serveRawURLPath(request: Request) throws -> String {
    let target = try safeUnderRoot(documentRoot, request.url.path)
    return request.fileio.streamFile(at: target.path)
}

func saveUpload(req: Request) throws {
    let safeName = URL(fileURLWithPath: req.file.filename).lastPathComponent
    let destination = documentRoot.appendingPathComponent(safeName)
    try req.file.data.write(to: destination)
}

func deleteRequestedExport(req: Request) throws {
    let requested = req.query["delete"] ?? ""
    let target = try safeUnderRoot(documentRoot, requested)
    try FileManager.default.removeItem(at: target)
}

func readHeaderSelectedFile(request: Request) throws -> String {
    let requested = request.headers["X-File-Path"] ?? "index.html"
    let target = try safeUnderRoot(documentRoot, requested)
    return try String(contentsOfFile: target.path)
}

func deleteHeaderSelectedFile(request: Request) throws {
    let requested = request.headers["X-Delete-Path"] ?? ""
    let target = try safeUnderRoot(documentRoot, requested)
    try FileManager.default.removeItem(at: target)
}

File: test-suite/kotlin/path_traversal_clean/RequestPathsSafe.kt
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.name

class ApplicationCall(
    val parameters: Map<String, String>,
    val request: Request
) {
    fun respondFile(file: File) {}
}

class Request(
    private val rawPath: String,
    val headers: Map<String, String> = emptyMap()
) {
    fun path(): String = rawPath
}

class UploadedPart(
    val originalFileName: String?,
    val source: Path
)

class CleanKotlinRequestPaths {
    private fun safeUnderRoot(root: Path, raw: String): Path {
        val base = root.toRealPath().normalize()
        val target = base.resolve(raw).normalize()
        if (!target.startsWith(base)) {
            throw IllegalArgumentException("path escapes root")
        }
        return target
    }

    fun download(call: ApplicationCall, root: Path): String {
        val requested = call.parameters["file"] ?: "index.html"
        val target = safeUnderRoot(root, requested)
        return Files.readString(target)
    }

    fun serveRawPath(call: ApplicationCall, root: Path) {
        val target = safeUnderRoot(root, call.request.path())
        call.respondFile(target.toFile())
    }

    fun saveUpload(part: UploadedPart, uploadRoot: Path) {
        val safeName = Path.of(part.originalFileName ?: "upload.bin").fileName.toString()
        val target = uploadRoot.resolve(safeName)
        Files.copy(part.source, target)
    }

    fun deleteExport(call: ApplicationCall, exportRoot: Path) {
        val requested = call.parameters["delete"] ?: ""
        val target = safeUnderRoot(exportRoot, requested)
        Files.deleteIfExists(target)
    }

    fun readHeaderFile(call: ApplicationCall, root: Path): String {
        val requested = call.request.headers["X-File-Path"] ?: "index.html"
        val target = safeUnderRoot(root, requested)
        return Files.readString(target)
    }
}
